normalize.tf

```diff
@@ -0,0 +1,146 @@
+# In this file, we normalize all the rules into full objects with all keys.
+# Then we partition the normalized rules for use as either inline or resourced rules.
+
+locals {
+
+  # Note: we have to use [] instead of null for unset lists due to
+  # https://github.com/hashicorp/terraform/issues/28137
+  # which was not fixed until Terraform 1.0.0
+  norm_rules = local.enabled ? [for rule in var.rules : {
+    type        = rule.type
+    from_port   = rule.from_port
+    to_port     = rule.to_port
+    protocol    = rule.protocol
+    description = lookup(rule, "description", local.default_rule_description)
+
+    # Convert a missing key, a value of null, or a value of empty list to []
+    cidr_blocks      = try(length(rule.cidr_blocks), 0) > 0 ? rule["cidr_blocks"] : []
```

Comment:
Interesting observation, @max-lobur. No magic, just an artifact. The current style guidelines permit either form, but `rule["cidr_blocks"]` is slightly preferred because the index in that case does not need to be a valid identifier. But `length(rule.cidr_blocks)` evolved from `length(lookup(rule, "cidr_blocks", ""))` and so the dot form more naturally followed. I suppose I should normalize them, but I honestly cannot decide which format I prefer.

Reply:
dot :)